File: c/complex-numbers/src/complex_numbers.c
#include "complex_numbers.h"
#include <math.h>

complex_t c_add(complex_t a, complex_t b) {
    return (complex_t) {
        .real = a.real + b.real,
        .imag = a.imag + b.imag,
    };
}

complex_t c_sub(complex_t a, complex_t b) {
    return (complex_t) {
        .real = a.real - b.real,
        .imag = a.imag - b.imag,
    };
}

complex_t c_mul(complex_t a, complex_t b) {
    return (complex_t) {
        .real = (a.real * b.real) - (a.imag * b.imag),
        .imag = (a.real * b.imag) + (a.imag * b.real),
    };
}

complex_t c_div(complex_t a, complex_t b) {
    complex_t conjugate = c_conjugate(b);
    complex_t dividend = c_mul(a, conjugate);
    complex_t divisor = c_mul(b, conjugate);
    return (complex_t) {
        .real = dividend.real / divisor.real,
        .imag = dividend.imag / divisor.real,
    };
}

double c_abs(complex_t x) {
    return sqrt(pow(x.real, 2) + pow(x.imag, 2));
}

complex_t c_conjugate(complex_t x) {
    return (complex_t) {
        .real = x.real,
        .imag = -1 * x.imag,
    };
}

double c_real(complex_t x) {
    return x.real;
}

double c_imag(complex_t x) {
    return x.imag;
}

complex_t c_exp(complex_t x) {
    return (complex_t) {
        .real = exp(x.real) * cos(x.imag),
        .imag = exp(x.real) * sin(x.imag),
    };
}

File: python/complex-numbers/complex_numbers.py
from math import exp, cos, sin

class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __repr__(self):
        return "{} + {}i".format(self.real, self.imaginary)

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        return ComplexNumber(
            self.real + other.real,
            self.imaginary + other.imaginary
        )

    def __mul__(self, other):
        return ComplexNumber(
            self.real * other.real - self.imaginary * other.imaginary,
            self.real * other.imaginary + self.imaginary * other.real
        )

    def __sub__(self, other):
        return ComplexNumber(
            self.real - other.real,
            self.imaginary - other.imaginary
        )

    def __truediv__(self, other):
        conjugate = other.conjugate()
        dividend = self * conjugate
        divisor = other * conjugate
        return ComplexNumber(
            dividend.real / divisor.real,
            dividend.imaginary / divisor.real
        )

    def __abs__(self):
        return (self.real ** 2 + self.imaginary ** 2) ** (1 / 2)

    def conjugate(self):
        return ComplexNumber(
            self.real,
            -1 * self.imaginary
        )

    def exp(self):
        return ComplexNumber(
            exp(self.real) * cos(self.imaginary),
            exp(self.real) * sin(self.imaginary)
        )
